botocore/docs/bcdoc/docstringparser.py

```diff
@@ -22,30 +21,175 @@ class DocStringParser(html_parser.HTMLParser):
     """
 
     def __init__(self, doc):
-        html_parser.HTMLParser.__init__(self)
+        six.moves.html_parser.HTMLParser.__init__(self)
         self.doc = doc
-        self.unhandled_tags = []
+        self.tree = HTMLTree(self.doc)
 
     def handle_starttag(self, tag, attrs):
-        handler_name = 'start_%s' % tag
-        if hasattr(self.doc.style, handler_name):
-            getattr(self.doc.style, handler_name)(attrs)
-        else:
-            self.unhandled_tags.append(tag)
+        self.tree.add_tag(tag, attrs=attrs)
 
     def handle_endtag(self, tag):
-        handler_name = 'end_%s' % tag
-        if hasattr(self.doc.style, handler_name):
-            getattr(self.doc.style, handler_name)()
+        self.tree.add_tag(tag, is_start=False)
 
     def handle_data(self, data):
-        if data.isspace():
-            data = ' '
+        self.tree.add_data(data)
+
+    def close(self):
+        # HTMLParser is an old style class, so the super() method will not work.
+        six.moves.html_parser.HTMLParser.close(self)
+        self.tree.write()
+        self.tree = None
```

Comment:
What happens if I tried to feed data into the parse after closing it? Would setting it to ``None`` cause attribute errors? I wonder if you should add a reset() method or just create a new HTMLTree after closing it.